smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/HttpQueryParamsTraitValidator.java

```diff
@@ -0,0 +1,82 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.validation.validators;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.HttpQueryParamsTrait;
+import software.amazon.smithy.model.traits.HttpQueryTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * When the `httpQueryParams` trait is used, this validator emits a NOTE when another member of the container shape
+ * applies the `httpQuery` trait which may result in a conflict within the query string.
+ */
+public final class HttpQueryParamsTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        if (!model.isTraitApplied(HttpQueryParamsTrait.class)) {
+            return Collections.emptyList();
+        } else {
+            return validateQueryTraitUsage(model);
+        }
+    }
+
+    private List<ValidationEvent> validateQueryTraitUsage(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        for (Shape shape : model.getShapesWithTrait(HttpQueryParamsTrait.class)) {
+            shape.asMemberShape().flatMap(member -> model.getShape(member.getContainer())
+                .flatMap(Shape::asStructureShape))
+                .ifPresent(structure -> {
+                    // Gather the names of member shapes, as strings, that apply HttpQuery or HttpQueryParams traits
+                    List<String> queryShapes = getMembersWithTrait(structure, HttpQueryTrait.class);
+                    List<String> queryParamsShapes = getMembersWithTrait(structure, HttpQueryParamsTrait.class);
+                    // Only one HttpQueryParams trait can be applied per structure.  If it isn't applied, no note will
+                    // be emitted.
+                    if (queryShapes.size() > 0 && queryParamsShapes.size() == 1) {
+                        events.add(createNote(structure, queryParamsShapes.get(0), queryShapes));
+                    }
+                });
+        }
+
+        return events;
+    }
+
+    private List<String> getMembersWithTrait(StructureShape structure, Class<? extends Trait> trait) {
+        List<String> members = new ArrayList<>();
+        for (MemberShape member : structure.members()) {
+            if (member.hasTrait(trait)) {
+                members.add(member.toShapeId().toString());
```

Comment:
Maybe just supply the name through `getMemberName`.

Reply:
Fixed in 7004468.